db/test/Test/Pos/DB/Block/Logic/SplitByEpoch.hs

```diff
@@ -0,0 +1,110 @@
+-- Generators and helpers
+
+-- For each element of a list of 'OldestFirst NE a' convert each sublist
+-- into a list of unique epoch indices.
+convertEpochs :: HasEpochIndex a => [OldestFirst NE a] -> [[Word64]]
+convertEpochs =
+    List.map (map (getEpochIndex . view epochIndexL) . NE.toList . getOldestFirst)
+
+wrapRight :: Word64 -> Either a Wrapper
+wrapRight = Right . Wrapper . EpochIndex
+
+genWrapperList :: Int -> Gen (OldestFirst NE Wrapper)
+genWrapperList shortestChunk = do
+    start <- Gen.word64 $ Range.constant 0 1000
+    end <- Gen.word64 $ Range.constant (start + 1) (start + 15)
+    OldestFirst . NE.fromList <$> concatMapM genSubList [start .. end]
+  where
+    genSubList :: Word64 -> Gen [Wrapper]
+    genSubList n =
+        fmap (Wrapper . EpochIndex) <$> Gen.list (Range.linear shortestChunk 10) (pure n)
+
+withTestCount :: TestLimit -> PropertyT IO () -> Property
+withTestCount n =
+    H.withTests n . H.property
+
+-- -----------------------------------------------------------------------------
+-- Tests
+
+-- Splitting followed by concatenation should produce the original list.
+prop_splitByEpoch_preserves_elements :: Property
+prop_splitByEpoch_preserves_elements =
+    withTestCount 1000 $ do
+        xs <- H.forAll $ genWrapperList 1
+        splitFirst <- H.forAll Gen.bool
+        concatEpochs (splitByEpoch splitFirst xs) === Just xs
+
+-- If each chunks is converted a list of 'Word64' and then `List.nub' is
+-- called on each chunk, then the length of each chunk should be 1.
+prop_all_chunks_are_same_epoch :: Property
+prop_all_chunks_are_same_epoch =
+    withTestCount 1000 $ do
+        xs <- H.forAll $ genWrapperList 1
+        H.assert $ List.all (\ys -> length ys == 1)
+                    . List.map List.nub
+                    $ convertEpochs (splitByEpoch True xs)
+
+-- Test the behavior of the 'splitFirst' parameter to 'splitByEpoch'.
+prop_spliting_first_is_longer :: Property
+prop_spliting_first_is_longer =
+    withTestCount 1000 $ do
+        xs <- H.forAll $ genWrapperList 1
+        let splitFirstTrue = splitByEpoch True xs
+            splitFirstFalse = splitByEpoch False xs
+        case splitFirstFalse of
+            [] -> H.failure -- Should never happen
+            (h:_) -> do
+                let initialSingleton = NE.length (getOldestFirst h) <= 1
+                if initialSingleton
+                    then length splitFirstTrue === length splitFirstFalse
+                    else length splitFirstTrue === 1 + length splitFirstFalse
+                H.cover 50 "non-trivial" (not initialSingleton)
+                H.cover 25 "trivial" initialSingleton
+
+prop_epoch_are_split :: Property
+prop_epoch_are_split =
+    withTestCount 1000 $ do
+        xs <- H.forAll $ genWrapperList 2
+        let epochCount = length . List.nub . map (view epochIndexL) . NE.toList $ getOldestFirst xs
+        length (splitByEpoch True xs) === 2 * epochCount
```

Comment:
Note to self / other reviewers:

`epochCount` is multiplied by 2 in L103 because `splitEpoch True` will chunk each epoch into 2 lists - the first block, and all the remaining blocks. Thus it'll double the length.